File: tools/svg_to_png.py
"""
svg_to_png.py — SVG → PNG 批量转换脚本
依赖：pip install cairosvg

用法：
  python tools/svg_to_png.py          # 全量转换
  python tools/svg_to_png.py --dry    # 仅预览，不写文件

规则：
  - bg_* 文件：原始 SVG 尺寸输出，宽度上限 2048px
  - 其余文件：4x 放大输出
  - frame_gold / frame_silver → 同时输出到 Assets/Resources/UI/（覆盖旧版）
  - 其余 → Assets/Resources/UI/Generated/
"""

import sys
import os
import re

try:
    import cairosvg
except ImportError:
    print("❌ 缺少依赖：cairosvg")
    print("   请运行：pip install cairosvg")
    sys.exit(1)

# ── 路径配置 ──────────────────────────────────────────────
SCRIPT_DIR   = os.path.dirname(os.path.abspath(__file__))
PROJECT_ROOT = os.path.dirname(SCRIPT_DIR)
SVG_DIR      = os.path.join(PROJECT_ROOT, "Assets", "Art", "SVG")
OUT_GENERATED = os.path.join(PROJECT_ROOT, "Assets", "Resources", "UI", "Generated")
OUT_UI        = os.path.join(PROJECT_ROOT, "Assets", "Resources", "UI")

# 需要额外输出到 Assets/Resources/UI/ 覆盖旧贴图的文件
OVERRIDE_FILES = {"frame_gold", "frame_silver"}

# bg_* 最大宽度（原尺寸，不 4x）
BG_MAX_WIDTH = 2048

# 默认缩放倍数
SCALE_4X = 4.0

# ── SVG 尺寸读取（正则，不依赖 lxml）────────────────────
_WH_RE = re.compile(r'<svg[^>]*\swidth="([0-9.]+)"[^>]*\sheight="([0-9.]+)"')

def get_svg_size(path: str):
    with open(path, "r", encoding="utf-8") as f:
        head = f.read(1024)
    m = _WH_RE.search(head)
    if m:
        return float(m.group(1)), float(m.group(2))
    return None, None

# ── 主转换逻辑 ────────────────────────────────────────────
def convert(svg_path: str, out_path: str, scale: float, dry: bool):
    w, h = get_svg_size(svg_path)
    if w and h:
        out_w = int(w * scale)
        out_h = int(h * scale)
    else:
        out_w = out_h = None

    if dry:
        size_str = f"{out_w}×{out_h}" if out_w else "auto"
        print(f"  [dry] {os.path.basename(svg_path)} → {os.path.relpath(out_path, PROJECT_ROOT)}  ({size_str})")
        return

    os.makedirs(os.path.dirname(out_path), exist_ok=True)
    kwargs = {"write_to": out_path}
    if out_w:
        kwargs["output_width"]  = out_w
        kwargs["output_height"] = out_h

    cairosvg.svg2png(url=svg_path, **kwargs)
    size_str = f"{out_w}×{out_h}" if out_w else "auto"
    print(f"  ✅ {os.path.basename(svg_path)} → {os.path.relpath(out_path, PROJECT_ROOT)}  ({size_str})")

def main():
    dry = "--dry" in sys.argv

    if not os.path.isdir(SVG_DIR):
        print(f"❌ SVG 目录不存在：{SVG_DIR}")
        sys.exit(1)

    svgs = sorted(f for f in os.listdir(SVG_DIR) if f.endswith(".svg"))
    if not svgs:
        print("⚠️  SVG 目录为空，无文件可转换")
        return

    print(f"{'[DRY RUN] ' if dry else ''}开始转换 {len(svgs)} 个 SVG → PNG\n")

    ok = 0
    errors = []

    for fname in svgs:
        stem = os.path.splitext(fname)[0]
        svg_path = os.path.join(SVG_DIR, fname)
        out_name = stem + ".png"

        # 决定输出路径
        out_path = os.path.join(OUT_GENERATED, out_name)

        # 决定缩放
        is_bg = stem.startswith("bg_")
        if is_bg:
            w, _ = get_svg_size(svg_path)
            scale = min(1.0, BG_MAX_WIDTH / w) if w else 1.0
        else:
            scale = SCALE_4X

        try:
            convert(svg_path, out_path, scale, dry)

            # frame_gold / frame_silver 同时覆盖 UI/ 目录
            if stem in OVERRIDE_FILES and not dry:
                override_path = os.path.join(OUT_UI, out_name)
                convert(svg_path, override_path, scale, dry)
                print(f"     ↳ 同步覆盖 Assets/Resources/UI/{out_name}")

            ok += 1
        except Exception as e:
            print(f"  ❌ {fname}: {e}")
            errors.append(fname)

    print(f"\n{'─'*50}")
    if dry:
        print(f"[DRY RUN] 共 {len(svgs)} 个文件，实际未写入")
    else:
        print(f"完成：{ok}/{len(svgs)} 成功", end="")
        if errors:
            print(f"，{len(errors)} 个失败：{errors}")
        else:
            print(" ✅")
        print(f"输出目录：{os.path.relpath(OUT_GENERATED, PROJECT_ROOT)}")

if __name__ == "__main__":
    main()

File: tools/svg_to_png.mjs
/**
 * svg_to_png.mjs — SVG → PNG 批量转换（Node.js + sharp）
 * 用法：node tools/svg_to_png.mjs [--dry]
 */

import { createRequire } from "module";
import path from "path";
import fs from "fs";
import { fileURLToPath } from "url";

const require = createRequire(import.meta.url);
const sharp = require("sharp");

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, "..");

const SVG_DIR       = path.join(ROOT, "Assets", "Art", "SVG");
const OUT_GENERATED = path.join(ROOT, "Assets", "Resources", "UI", "Generated");
const OUT_UI        = path.join(ROOT, "Assets", "Resources", "UI");

// frame_gold/silver 不再覆盖 Assets/Resources/UI/（原始透明帧文件保留）
const OVERRIDE_FILES = new Set([]);
const BG_MAX_WIDTH   = 2048;
const SCALE_4X       = 4;

const dry = process.argv.includes("--dry");

// 从 SVG 读取 width/height
function getSvgSize(svgPath) {
  const head = fs.readFileSync(svgPath, "utf8").slice(0, 1024);
  const m = head.match(/<svg[^>]*\swidth="([0-9.]+)"[^>]*\sheight="([0-9.]+)"/);
  if (m) return { w: parseFloat(m[1]), h: parseFloat(m[2]) };
  return { w: null, h: null };
}

async function convert(svgPath, outPath, scale) {
  const { w, h } = getSvgSize(svgPath);
  const outW = w ? Math.round(w * scale) : undefined;
  const outH = h ? Math.round(h * scale) : undefined;

  const label = `${path.basename(svgPath)} → ${path.relative(ROOT, outPath)}  (${outW ?? "auto"}×${outH ?? "auto"})`;

  if (dry) {
    console.log(`  [dry] ${label}`);
    return;
  }

  fs.mkdirSync(path.dirname(outPath), { recursive: true });
  const svgContent = fs.readFileSync(svgPath);
  let img = sharp(svgContent, { density: 192 });
  if (outW && outH) img = img.resize(outW, outH);
  await img.png({ compressionLevel: 9 }).toFile(outPath);
  console.log(`  ✅ ${label}`);
}

async function main() {
  if (!fs.existsSync(SVG_DIR)) {
    console.error(`❌ SVG 目录不存在：${SVG_DIR}`);
    process.exit(1);
  }

  const svgs = fs.readdirSync(SVG_DIR).filter(f => f.endsWith(".svg")).sort();
  if (!svgs.length) { console.log("⚠️  SVG 目录为空"); return; }

  console.log(`${dry ? "[DRY RUN] " : ""}开始转换 ${svgs.length} 个 SVG → PNG\n`);

  let ok = 0;
  const errors = [];

  for (const fname of svgs) {
    const stem    = path.basename(fname, ".svg");
    const svgPath = path.join(SVG_DIR, fname);
    const outName = stem + ".png";
    const outPath = path.join(OUT_GENERATED, outName);

    // 背景类：原尺寸（最大 2048），其余 4x
    const isBg = stem.startsWith("bg_");
    let scale = SCALE_4X;
    if (isBg) {
      const { w } = getSvgSize(svgPath);
      scale = w ? Math.min(1, BG_MAX_WIDTH / w) : 1;
    }

    try {
      await convert(svgPath, outPath, scale);

      // frame_gold / frame_silver 同时写到 Assets/Resources/UI/
      if (OVERRIDE_FILES.has(stem)) {
        const overridePath = path.join(OUT_UI, outName);
        await convert(svgPath, overridePath, scale);
        if (!dry) console.log(`     ↳ 覆盖 Assets/Resources/UI/${outName}`);
      }

      ok++;
    } catch (e) {
      console.error(`  ❌ ${fname}: ${e.message}`);
      errors.push(fname);
    }
  }

  console.log("\n" + "─".repeat(50));
  if (dry) {
    console.log(`[DRY RUN] ${svgs.length} 个文件，未实际写入`);
  } else {
    console.log(`完成：${ok}/${svgs.length} 成功${errors.length ? `，失败：${errors.join(", ")}` : " ✅"}`);
    console.log(`输出：${path.relative(ROOT, OUT_GENERATED)}`);
  }
}

main().catch(e => { console.error(e); process.exit(1); });
